src/thread_event.c

```diff
@@ -0,0 +1,254 @@
+	uint64_t accumbytes = thread_allocated_after -
+	    thread_allocated_last_event_get(tsd);
+
+	if (accumbytes < usize) {
+		/* accumbytes overflowed. */
+		accumbytes = UINT64_MAX;
+	}
+
+	uint64_t wait = THREAD_EVENT_MAX_START_WAIT;
+
+#define E(event, condition)						\
+	if (condition) {						\
+		uint64_t event_wait = event##_event_wait_get(tsd);	\
+		assert(event_wait <= THREAD_EVENT_MAX_START_WAIT);	\
+		if (event_wait > accumbytes) {				\
+			event_wait -= accumbytes;			\
+		} else {						\
+			event_wait = 0U;				\
+			if (delay_event || !tsd_nominal(tsd) ||		\
+			    tsd_reentrancy_level_get(tsd) > 0) {	\
+				event_wait =				\
+				    THREAD_EVENT_MIN_START_WAIT;	\
+			}						\
+		}							\
+		assert(event_wait <= THREAD_EVENT_MAX_START_WAIT);	\
+		event##_event_wait_set(tsd, event_wait);		\
+		/*							\
+		 * If there is a single event, then the remaining wait	\
+		 * time may become zero, and we rely on either the	\
+		 * event handler or a thread_event_update() call later	\
+		 * to properly set next_event; if there are multiple	\
+		 * events, then	here we can get the minimum remaining	\
+		 * wait time to	the next already set event.		\
+		 */							\
+		if (event_wait > 0U && event_wait < wait) {		\
+			wait = event_wait;				\
+		}							\
+	}
+
+	ITERATE_OVER_ALL_EVENTS
+#undef E
+
+	assert(wait <= THREAD_EVENT_MAX_START_WAIT);
+	thread_allocated_last_event_set(tsd, thread_allocated_after);
+	thread_event_adjust_thresholds_helper(tsd, thread_allocated_after,
+	    wait);
+
+	thread_event_assert_invariants(tsd);
+
+#define E(event, condition)						\
+	if (condition && event##_event_wait_get(tsd) == 0U) {		\
+		assert(!delay_event);					\
+		assert(tsd_nominal(tsd));				\
+		assert(tsd_reentrancy_level_get(tsd) == 0);		\
+		thread_##event##_event_handler(tsd);			\
+	}
+
+	ITERATE_OVER_ALL_EVENTS
+#undef E
+
+	thread_event_assert_invariants(tsd);
+}
+
+void
+thread_event_rollback(tsd_t *tsd, size_t diff) {
+	thread_event_assert_invariants(tsd);
+
+	if (diff == 0U) {
+		return;
+	}
+
+	uint64_t thread_allocated = thread_allocated_get(tsd);
+	/* The subtraction is intentionally susceptible to underflow. */
+	uint64_t thread_allocated_rollback = thread_allocated - diff;
+	thread_allocated_set(tsd, thread_allocated_rollback);
+
+	uint64_t last_event = thread_allocated_last_event_get(tsd);
+	/* Both subtractions are intentionally susceptible to underflow. */
+	if (thread_allocated_rollback - last_event <=
+	    thread_allocated - last_event) {
+		thread_event_assert_invariants(tsd);
+		return;
+	}
+
+	thread_allocated_last_event_set(tsd, thread_allocated_rollback);
+
+	/* The subtraction is intentionally susceptible to underflow. */
+	uint64_t wait_diff = last_event - thread_allocated_rollback;
+	assert(wait_diff <= diff);
+
+#define E(event, condition)						\
+	if (condition) {						\
+		uint64_t event_wait = event##_event_wait_get(tsd);	\
+		assert(event_wait <= THREAD_EVENT_MAX_START_WAIT);	\
+		if (event_wait > 0U) {					\
```

Comment:
Question: why this condition?

Reply:
Because I don't want to adjust the wait times for the already-triggered events. The reason is that there's no way to know how much exactly should we adjust them. We only know that they've been triggered, but we've lost their triggering points: we already lost the previous `next_event`, and even if we still have it, the event triggering point can be anywhere below the current `thread_allocated` and above the previous `next_event`.

Reply:
Makes sense; another question, if the event was triggered, and it got replenished already which means non zero here, then we are rolling back incorrectly? Is that a valid concern?

Reply:
To make it clear, if that’s an issue, I don’t mean we have to fix it right away. I believe the refactoring on the prof sample path will fix this for us, i.e. allow us to call thread_event after the allocation. Then the rollback would be purely for OOM and won’t concern us the same way.

Reply:
I think there are a few cases here:

(1) If an event was not triggered, then `thread_event_rollback()` won't trigger it and will increase the wait time so that the remaining wait time is correct with respect to the rolled-back `thread_allocated`.

(2) If an event was triggered but a new wait time is already set, then `thread_event_rollback()` still increases the wait time.

(3) If an event was triggered and a new wait time has not been set, then we'll leave the wait time in the "triggered" state and let subsequent logic properly set it (similar to how `thread_event_trigger()` works).

From correctness point of view, I think (2) and (3) each have their own advantages. The advantage of (3) is that the next wait time is correct. This is a disadvantage of (2), but meanwhile its advantage is that the *average* wait time actually becomes correct, considering the fact that the previous triggering over-counted. In other words, imagine if `thread_event_rollback()` takes place frequently, then (2) won't affect the average event frequency whereas (3) increases the frequency.

I guess no matter which way we prefer, it seems hard to make (2) work like (3) or vice versa. To make (2) work like (3), it's hard to tell whether a non-zero wait time was just set or not; to make (3) work like (2), it's hard to inform subsequent wait-time-setting logic to further increase the wait time by XXX.

I think for `xallocx()` case, the first rollback (inside `ixallocx_prof()`) follows (3) whereas the second rollback (inside `je_xallocx()`) follows (2). :)

Thoughts?

Reply:
I vote for the current approach. Like you mentioned it wouldn’t be easy to fully restore a previous state, and I don’t feel it’s worth the effort given we expect the refactoring. The difficulty here says again that being able to move thread_event to after the allocation is important.